tsl/src/continuous_aggs/options.c

```diff
@@ -218,28 +218,37 @@ cagg_get_compression_params(ContinuousAgg *agg, Hypertable *mat_ht)
 	return defelems;
 }
 
-/* enable/disable compression on continuous aggregate */
+/* forwards compression related changes via an alter statement to the underlying HT */
 static void
-cagg_alter_compression(ContinuousAgg *agg, Hypertable *mat_ht, bool compress_enable)
+cagg_alter_compression(ContinuousAgg *agg, Hypertable *mat_ht, List *compress_defelems)
 {
-	List *defelems = NIL;
 	Assert(mat_ht != NULL);
-	if (compress_enable)
-		defelems = cagg_get_compression_params(agg, mat_ht);
+	WithClauseResult *with_clause_options =
+		ts_compress_hypertable_set_clause_parse(compress_defelems);
 
-	DefElem *enable = makeDefElemExtended("timescaledb",
-										  "compress",
-										  compress_enable ? (Node *) makeString("true") :
-															(Node *) makeString("false"),
-										  DEFELEM_UNSPEC,
-										  -1);
-	defelems = lappend(defelems, enable);
+	if (with_clause_options[CompressEnabled].parsed)
+	{
+		List *default_compress_defelems = cagg_get_compression_params(agg, mat_ht);
+		WithClauseResult *default_with_clause_options =
+			ts_compress_hypertable_set_clause_parse(default_compress_defelems);
+		// merge defaults if there's any
+		for (int i = 0; i < _CompressOptionMax; i++)
+		{
+			if (with_clause_options[i].is_default && !default_with_clause_options[i].is_default)
+			{
+				with_clause_options[i] = default_with_clause_options[i];
+				// ereport(NOTICE,
+				// 		errmsg("defaulting %s to %s",
+				// 			   ts_compress_with_clause_def_get_name(i),
+				// 			   DatumGetCString(default_with_clause_options[i].parsed)));
```

Comment:
these `WithClauseDefinition` structs are are not really giving much advantage...
instead of enum keyed arrays - which could possibly wreck havoc if someone uses a different enum to index them - I would either add the key/type to the result as well; or try some other approach - like creating a full standard struct with the names of the fileds documenting what those informations are

without the name of the field its even hard to give proper messages to the user...

Reply:
I've decided to add a pointer for the definition to make unparsing/name access simple and straightforward